SalesforceCommerceCloud/pwa-kit
# @W15458525@ [App Extensibility ⚙️] Webpack Overrides Plugin

# Description

This PR builds on top of #1826 where we created the utility function required for resolving wildcard imports. In this work we take those utilities and put them together in the form of a webpack plugin called `OverridesResolverPlugin`. The purpose of this plugin is to allow any PWA-Kit application to opt into the override module resolution via the wildcard import syntax. 

What does that all mean? It means if you are defining an Application Extension and you want to open your extension up to _low-level_ modification via file replacement, all you have to do is use the wildcard import syntax. 

Here is an example:

```
// @salesforce/extension-sample/src/setup-app.ts
const SamplePage = loadable(() => import('*/pages/sample'))

class Sample extends ApplicationExtension {
   
    extendRoutes(routes: IRouteConfig[]): IRouteConfig[] {
        return [
            {
                exact: true,
                path: '/sample-page',
                component: SamplePage
            },
            ...routes
        ]
    }
}

export default Sample
```

Using the wildcard import syntax when importing in this example means that any Application Extension or the Base Application can override the `pages/sample.{ts|js|tsx|tsx}` file in their `overrides` folder.


# Types of Changes


> Breaking changes include:
>
> - Removing a public function or component or prop
> - Adding a required argument to a function
> - Changing the data type of a function parameter or return value
> - Adding a new peer dependency to `package.json`

# Changes

- Created `OverridesResolverPlugin` plugin
- Add tests for plugin
- Update old tests to account for the `overrides` folder (this was not something that was required previously)

# How to Test-Drive This PR

- The "base" project used to demo this feature (typescript-minimal) was deployed here --> https://scaffold-pwa-bens-playground.mobify-storefront.com/
- If you see the below UI (minus the annotations) than it is working for you, you can also run this locally.

![Screenshot 2024-09-17 at 4 01 56 PM](https://github.com/user-attachments/assets/b452cefb-41ab-4d7f-8fab-4b73379fcb1e)

## Proof it's working in windows


https://github.com/user-attachments/assets/665d9d56-9c2a-435e-b27e-e8f0e0c9916a


# Checklists





## General


## Accessibility Compliance

You must check off all items in **one** of the follow two lists:


_or..._


## Localization
